File: Data Structures & Algorithms/permutations-ii/submission-1.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        visited = set()

        def rec(l):
            if len(l) == len(nums):
                res.append(l[:])
                return

            for i in range(len(nums)):
                if i > 0 and nums[i] == nums[i-1] and i-1 not in visited:
                    continue
                if i in visited:
                    continue
                visited.add(i)
                l.append(nums[i])
                rec(l)
                visited.remove(i)
                l.pop()
        rec([])
        return res

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int prefix = 1;
        int[] res = new int[nums.length];
        for (int i = 0; i < nums.length; i++) {
            res[i] = 1;
        }
        for (int i = 0; i < nums.length; i++) {
            res[i] *= prefix;
            prefix *= nums[i];
        }
        int posfix = 1;
        for (int i = nums.length - 1; i >= 0; i--) {
            res[i] *= posfix;
            posfix *= nums[i];
        }
        return res;
    }
}
